k8s/apiserver/apiserver.go

```diff
@@ -0,0 +1,174 @@
+package apiserver
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/version"
+	genericregistry "k8s.io/apiserver/pkg/registry/generic"
+	"k8s.io/apiserver/pkg/registry/rest"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+
+	"github.com/grafana/grafana-app-sdk/app"
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+type APIServerInstaller struct {
+	AppProvider         app.Provider
+	AppConfig           app.Config
+	ManagedKindResolver func(kind, version string) (resource.Kind, error)
+
+	app    app.App
+	scheme *runtime.Scheme
+	codecs serializer.CodecFactory
+}
+
+func (r *APIServerInstaller) AddToScheme(scheme *runtime.Scheme) error {
+	kindsByGV, err := r.getKindsByGroupVersion()
+	if err != nil {
+		return fmt.Errorf("failed to get kinds by group version: %w", err)
+	}
+
+	internalKinds := map[string]resource.Kind{}
+	kindsByGroup := map[string][]resource.Kind{}
+	groupVersions := []schema.GroupVersion{}
+	for gv, kinds := range kindsByGV {
+		for _, kind := range kinds {
+			scheme.AddKnownTypeWithName(kind.GroupVersionKind(), kind.ZeroValue())
+			scheme.AddKnownTypeWithName(gv.WithKind(kind.Kind()+"List"), kind.ZeroListValue())
+			metav1.AddToGroupVersion(scheme, kind.GroupVersionKind().GroupVersion())
+			if _, ok := internalKinds[kind.Kind()]; !ok {
+				internalKinds[kind.Kind()] = kind
+			}
+			if _, ok := kindsByGroup[kind.Group()]; !ok {
+				kindsByGroup[kind.Group()] = []resource.Kind{}
+			}
+			kindsByGroup[kind.Group()] = append(kindsByGroup[kind.Group()], kind)
+		}
+		groupVersions = append(groupVersions, gv)
+	}
+
+	internalGv := schema.GroupVersion{Group: r.AppConfig.ManifestData.Group, Version: runtime.APIVersionInternal}
+	for _, internalKind := range internalKinds {
+		scheme.AddKnownTypeWithName(internalGv.WithKind(internalKind.Kind()), internalKind.ZeroValue())
+		scheme.AddKnownTypeWithName(internalGv.WithKind(internalKind.Kind()+"List"), internalKind.ZeroListValue())
+
+		for _, kind := range kindsByGroup[internalKind.Group()] {
+			scheme.AddConversionFunc(kind.ZeroValue(), internalKind.ZeroValue(), r.conversionHandler)
+			scheme.AddConversionFunc(internalKind.ZeroValue(), kind.ZeroValue(), r.conversionHandler)
```

Comment:
@IfSentient i had started restructuring this when i put it in my git stash, but i just added back roughly what i was thinking for conversion